File: src/c/prime_counter.c
#include <stdio.h>
#include <time.h>
#include <stdbool.h>
#include "prime_counter.h"

#define RUNS 4
#define AMOUNT 40000

#define NOT_DEBUGGING
int primes[AMOUNT];

#ifdef NOT_DEBUGGING
int main() {
    clock_t tests[RUNS];
    int array_size = 0;

    for (int i = 0; i < RUNS; i++) {
        clock_t timer = clock();
        counter();
        timer = clock() - timer;
        tests[i] = timer;
        array_size++;
    }

    float total = 0;
    for (int i = 0; i < RUNS; i++) {
        total += (double)tests[i] / CLOCKS_PER_SEC;
        printf("Runs in: %.3f seconds\n", (float)tests[i] / CLOCKS_PER_SEC);
    }

    double avg = total / array_size;
    printf("Took: %d tests. Average: %.3f seconds", RUNS, avg);
    return 0;
}
#endif

int counter(void) {
    int found = 0;
    int number = 2;
    int array_size = 0;

    while (found < AMOUNT) {
        bool flag_break = false;
        for (int i = 0; i < array_size; i++) {
            if (number % primes[i] == 0) {
                flag_break = true;
                break;
            }
        }

        if (!flag_break) {
            primes[found] = number;
            found++;
            array_size++;
            flag_break = false;
        }

        number++;
    }
    
    printf("%d\n", primes[0]);
    return array_size;
}

File: src/python/prime_counter.py
import time
from typing import Final

RUNS: Final[int] = 10
AMOUNT: int = 20000 # Not really constant because the test module may modify it?
PRIME_AMOUNT: Final[bool] = True


def prime_counter() -> list[int]:
    primes: list[int] = []
    primes_found: int = 0
    number: int = 2

    while (primes_found if PRIME_AMOUNT else number) < AMOUNT:
        for i in primes:
            if number % i == 0:
                break
        else:
            primes.append(number)
            primes_found += 1

        number += 1

    return primes


def main() -> None:
    if AMOUNT <= 0 or RUNS <= 0:
        print("AMOUNT or RUNS must be greater than 0!")
        return
    
    print("Prime Counting Benchmark: Python")
    total_time: int = 0
    primes: list[int] = []

    for run in range(RUNS):
        interval: float = time.perf_counter()
        primes = prime_counter()
        interval = time.perf_counter() - interval
        total_time += interval
        print(f"- Test {run + 1} ran in: {interval:.3f} seconds")

    print(f"\nCounted {len(primes)}"
    f"{'' if PRIME_AMOUNT else f' up to {AMOUNT}'}"
    f" in {RUNS} with an average of {(total_time/ RUNS):.3f} seconds"
    )


if __name__ == "__main__":
    main()